File: rxhttp-compiler/src/main/java/com/rxhttp/compiler/ParserAnnotatedClass.java
package com.rxhttp.compiler;


import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.MethodSpec.Builder;
import com.squareup.javapoet.ParameterSpec;
import com.squareup.javapoet.ParameterizedTypeName;
import com.squareup.javapoet.TypeName;
import com.squareup.javapoet.TypeVariableName;

import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

import javax.annotation.processing.Filer;
import javax.lang.model.element.Element;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.element.TypeParameterElement;
import javax.lang.model.type.MirroredTypesException;
import javax.lang.model.type.TypeMirror;

import rxhttp.wrapper.annotation.Parser;


public class ParserAnnotatedClass {

    private Map<String, TypeElement> mElementMap;
    private Map<String, List<? extends TypeMirror>> mTypeMap;

    public ParserAnnotatedClass() {
        mElementMap = new LinkedHashMap<>();
        mTypeMap = new LinkedHashMap<>();
    }

    public void add(TypeElement typeElement) {
        Parser annotation = typeElement.getAnnotation(Parser.class);
        String name = annotation.name();
        if (name.length() == 0) {
            throw new IllegalArgumentException(
                String.format("methodName() in @%s for class %s is null or empty! that's not allowed",
                    Parser.class.getSimpleName(), typeElement.getQualifiedName().toString()));
        }
        try {
            annotation.wrappers();
        } catch (MirroredTypesException e) {
            List<? extends TypeMirror> typeMirrors = e.getTypeMirrors();
            mTypeMap.put(name, typeMirrors);
        }
        mElementMap.put(name, typeElement);
    }

    public List<MethodSpec> getMethodList(Filer filer) {
        TypeVariableName t = TypeVariableName.get("T");
        TypeVariableName k = TypeVariableName.get("K");
        TypeVariableName v = TypeVariableName.get("V");
        ClassName callName = ClassName.get("okhttp3", "Call");
        ClassName okHttpClientName = ClassName.get("okhttp3", "OkHttpClient");
        ClassName responseName = ClassName.get("okhttp3", "Response");
        ClassName schedulerName = ClassName.get("io.reactivex", "Scheduler");
        ClassName observableName = ClassName.get("io.reactivex", "Observable");
        ClassName consumerName = ClassName.get("io.reactivex.functions", "Consumer");

        ClassName bitmapName = ClassName.get("android.graphics", "Bitmap");
        ClassName okResponseName = ClassName.get("okhttp3", "Response");
        ClassName headersName = ClassName.get("okhttp3", "Headers");
        ClassName httpSenderName = ClassName.get("rxhttp", "HttpSender");
        ClassName requestName = ClassName.get("okhttp3", "Request");
        ClassName parserName = ClassName.get("rxhttp.wrapper.parse", "Parser");
        ClassName progressName = ClassName.get("rxhttp.wrapper.entity", "Progress");
        ClassName progressTName = ClassName.get("rxhttp.wrapper.entity", "ProgressT");
        ClassName downloadParserName = ClassName.get("rxhttp.wrapper.parse", "DownloadParser");

        TypeName typeName = TypeName.get(String.class);
        TypeName listTName = ParameterizedTypeName.get(ClassName.get(List.class), t);
        TypeName mapTTName = ParameterizedTypeName.get(ClassName.get(Map.class), t, t);
        TypeName mapKVName = ParameterizedTypeName.get(ClassName.get(Map.class), k, v);
        TypeName progressTStringName = ParameterizedTypeName.get(progressTName, typeName);
        TypeName observableTName = ParameterizedTypeName.get(observableName, t);
        TypeName observableStringName = ParameterizedTypeName.get(observableName, typeName);
        TypeName consumerProgressName = ParameterizedTypeName.get(consumerName, progressName);
        TypeName parserTName = ParameterizedTypeName.get(parserName, t);

        List<MethodSpec> methodList = new ArrayList<>();
        MethodSpec.Builder method;
        method = MethodSpec.methodBuilder("execute")
            .addModifiers(Modifier.PUBLIC)
            .addException(IOException.class)
            .addStatement("doOnStart()")
            .addStatement("return $T.execute(param)", httpSenderName)
            .returns(responseName);
        methodList.add(method.build());

        method = MethodSpec.methodBuilder("execute")
            .addModifiers(Modifier.PUBLIC)
            .addTypeVariable(t)
            .addException(IOException.class)
            .addParameter(parserTName, "parser")
            .addStatement("return parser.onParse(execute())", httpSenderName)
            .returns(t);
        methodList.add(method.build());

        methodList.add(
            MethodSpec.methodBuilder("newCall")
                .addModifiers(Modifier.PUBLIC)
                .addStatement("return newCall(getOkHttpClient())")
                .returns(callName)
                .build());

        methodList.add(
            MethodSpec.methodBuilder("newCall")
                .addModifiers(Modifier.PUBLIC)
                .addParameter(okHttpClientName, "okHttp")
                .addStatement("return $T.newCall(okHttp, buildRequest())", httpSenderName)
                .returns(callName)
                .build());

        methodList.add(
            MethodSpec.methodBuilder("buildRequest")
                .addAnnotation(Override.class)
                .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
                .addStatement("doOnStart()")
                .addStatement("return param.buildRequest()")
                .returns(requestName)
                .build());

        methodList.add(
            MethodSpec.methodBuilder("doOnStart")
                .addJavadoc("请求开始前内部调用，用于添加默认域名等操作\n")
                .addStatement("setConverter(param)")
                .addStatement("addDefaultDomainIfAbsent(param)")
                .build());

        method = MethodSpec.methodBuilder("subscribeOn")
            .addModifiers(Modifier.PUBLIC)
            .addParameter(schedulerName, "scheduler")
            .addStatement("this.scheduler=scheduler")
            .addStatement("return (R)this")
            .returns(RxHttpGenerator.r);
        methodList.add(method.build());

        method = MethodSpec.methodBuilder("subscribeOnCurrent")
            .addJavadoc("设置在当前线程发请求\n")
            .addModifiers(Modifier.PUBLIC)
            .addStatement("this.scheduler=null")
            .addStatement("return (R)this")
            .returns(RxHttpGenerator.r);
        methodList.add(method.build());

        method = MethodSpec.methodBuilder("subscribeOnIo")
            .addModifiers(Modifier.PUBLIC)
            .addStatement("this.scheduler=Schedulers.io()")
            .addStatement("return (R)this")
            .returns(RxHttpGenerator.r);
        methodList.add(method.build());

        method = MethodSpec.methodBuilder("subscribeOnComputation")
            .addModifiers(Modifier.PUBLIC)
            .addStatement("this.scheduler=Schedulers.computation()")
            .addStatement("return (R)this")
            .returns(RxHttpGenerator.r);
        methodList.add(method.build());

        method = MethodSpec.methodBuilder("subscribeOnNewThread")
            .addModifiers(Modifier.PUBLIC)
            .addStatement("this.scheduler=Schedulers.newThread()")
            .addStatement("return (R)this")
            .returns(RxHttpGenerator.r);
        methodList.add(method.build());

        method = MethodSpec.methodBuilder("subscribeOnSingle")
            .addModifiers(Modifier.PUBLIC)
            .addStatement("this.scheduler=Schedulers.single()")
            .addStatement("return (R)this")
            .returns(RxHttpGenerator.r);
        methodList.add(method.build());

        method = MethodSpec.methodBuilder("subscribeOnTrampoline")
            .addModifiers(Modifier.PUBLIC)
            .addStatement("this.scheduler=Schedulers.trampoline()")
            .addStatement("return (R)this")
            .returns(RxHttpGenerator.r);
        methodList.add(method.build());

        method = MethodSpec.methodBuilder("asParser")
            .addAnnotation(Override.class)
            .addModifiers(Modifier.PUBLIC)
            .addTypeVariable(t)
            .addParameter(parserTName, "parser")
            .addStatement("doOnStart()")
            .addStatement("Observable<T> observable=$T.syncFrom(param,parser)", httpSenderName)
            .beginControlFlow("if(scheduler!=null)")
            .addStatement("observable=observable.subscribeOn(scheduler)")
            .endControlFlow()
            .addStatement("return observable")
            .returns(observableTName);
        methodList.add(method.build());

        RxHttpExtensions rxHttpExtensions = new RxHttpExtensions();

        //获取自定义的解析器
        for (Entry<String, TypeElement> item : mElementMap.entrySet()) {
            TypeMirror returnType = null; //获取onParse方法的返回类型
            TypeElement typeElement = item.getValue();
            for (Element element : typeElement.getEnclosedElements()) {
                if (!(element instanceof ExecutableElement)
                    || !element.getModifiers().contains(Modifier.PUBLIC)
                    || element.getModifiers().contains(Modifier.STATIC))
                    continue;
                ExecutableElement executableElement = (ExecutableElement) element;
                if (executableElement.getSimpleName().toString().equals("onParse")
                    && executableElement.getParameters().size() == 1
                    && executableElement.getParameters().get(0).asType().toString().equals("okhttp3.Response")) {
                    returnType = executableElement.getReturnType();
                    break;
                }
            }
            if (returnType == null) continue;
            String parserAlias = item.getKey();  //Parser注解里面的name字段值
            rxHttpExtensions.generateAsClassFun(typeElement, parserAlias);

            methodList.add(generateAsXxxMethod(typeElement, parserAlias, returnType, null));
            List<? extends TypeMirror> typeMirrors = mTypeMap.get(parserAlias);
            for (TypeMirror mirror : typeMirrors) {  //遍历Parser注解里面的wrappers数组
                String name = mirror.toString();
                String simpleName = name.substring(name.lastIndexOf(".") + 1);
                String methodName = parserAlias + simpleName;
                methodList.add(generateAsXxxMethod(typeElement, methodName, returnType, mirror));
            }
        }
        rxHttpExtensions.generateClassFile(filer);

        method = MethodSpec.methodBuilder("asDownload")
            .addAnnotation(Override.class)
            .addModifiers(Modifier.PUBLIC)
            .addParameter(String.class, "destPath")
            .addParameter(consumerProgressName, "progressConsumer")
            .addParameter(schedulerName, "observeOnScheduler")
            .addStatement("doOnStart()")
            .addStatement("Observable<Progress> observable = $T.downloadProgress(param, destPath, breakDownloadOffSize, scheduler)", httpSenderName)
            .beginControlFlow("if(observeOnScheduler != null)")
            .addStatement("observable = observable.observeOn(observeOnScheduler)")
            .endControlFlow()
            .addStatement("return observable.doOnNext(progressConsumer)\n" +
                ".filter(progress -> progress instanceof ProgressT)\n" +
                ".map(progress -> (($T) progress).getResult())", progressTStringName)
            .returns(observableStringName);
        methodList.add(method.build());

        return methodList;
    }

    private MethodSpec generateAsXxxMethod(
        TypeElement typeElement,
        String methodName,
        TypeMirror returnTypeMirror,
        @Nullable TypeMirror mirror
    ) {
        List<TypeVariableName> typeVariableNames = new ArrayList<>();
        List<ParameterSpec> parameterSpecs = new ArrayList<>();
        List<? extends TypeParameterElement> typeParameters = typeElement.getTypeParameters();
        for (TypeParameterElement element : typeParameters) {
            TypeVariableName typeVariableName = TypeVariableName.get(element);
            typeVariableNames.add(typeVariableName);
            ParameterSpec parameterSpec = ParameterSpec.builder(
                ParameterizedTypeName.get(ClassName.get(Class.class), typeVariableName),
                element.asType().toString().toLowerCase() + "Type").build();
            parameterSpecs.add(parameterSpec);
        }

        //自定义解析器对应的asXxx方法里面的语句
        StringBuilder statementBuilder = new StringBuilder("return asParser(new $T");
        int size = typeVariableNames.size();
        if (size > 0) statementBuilder.append("<");
        for (int i = 0; i < size; i++) {//添加泛型
            if (mirror != null) {
                String name = mirror.toString();
                String simpleName = name.substring(name.lastIndexOf('.') + 1);
                statementBuilder.append(simpleName).append("<");
            }
            TypeVariableName variableName = typeVariableNames.get(i);
            statementBuilder.append(variableName.name);
            if (mirror != null) {
                statementBuilder.append(">");
            }
            statementBuilder.append(i == size - 1 ? ">" : ",");
        }
        if (mirror != null) {
            String name = mirror.toString();
            String simpleName = name.substring(name.lastIndexOf('.') + 1);
            statementBuilder.append("(");
            for (ParameterSpec spec : parameterSpecs) {
                statementBuilder.append(spec.name).append(simpleName)
                    .append(",");
            }
            statementBuilder.deleteCharAt(statementBuilder.length() - 1).append(")");
        } else {
            statementBuilder.append("(");
            size = parameterSpecs.size();
            for (int i = 0; i < size; i++) {//添加参数
                ParameterSpec parameterSpec = parameterSpecs.get(i);
                statementBuilder.append(parameterSpec.name);
                if (i < size - 1) {
                    statementBuilder.append(",");
                }
            }
            statementBuilder.append(")");
        }
        statementBuilder.append(")");

        TypeName typeName = TypeName.get(returnTypeMirror);
        if (mirror != null) {
            ClassName className = ClassName.bestGuess(mirror.toString());
            if (typeName instanceof ParameterizedTypeName) {
                ParameterizedTypeName parameterizedTypeName = (ParameterizedTypeName) typeName;
                List<TypeName> typeNames = new ArrayList<>();
                for (TypeName type : parameterizedTypeName.typeArguments) {
                    TypeName parameterizedReturnType = ParameterizedTypeName
                        .get(className, type);
                    typeNames.add(parameterizedReturnType);
                }
                typeName = ParameterizedTypeName.get(parameterizedTypeName.rawType,
                    typeNames.toArray(new TypeName[0]));
            } else {
                typeName = ParameterizedTypeName.get(className, typeName);
            }
        }
        ClassName observableName = ClassName.get("io.reactivex", "Observable");
        TypeName returnType = ParameterizedTypeName.get(observableName, typeName);

        Builder builder = MethodSpec.methodBuilder("as" + methodName)
            .addModifiers(Modifier.PUBLIC);

        if (mirror != null) {
            ClassName parameterizedType = ClassName.get("rxhttp.wrapper.entity", "ParameterizedTypeImpl");
            ClassName type = ClassName.get("java.lang.reflect", "Type");
            for (ParameterSpec spec : parameterSpecs) {
                String expression = "$T " + spec.name + "$T = $T.get($T.class, " + spec.name + ")";
                builder.addStatement(expression, type, TypeName.get(mirror), parameterizedType, TypeName.get(mirror));
            }
        }

        builder.addTypeVariables(typeVariableNames)  //添加泛型
            .addParameters(parameterSpecs)     //添加参数
            .addStatement(statementBuilder.toString(), ClassName.get(typeElement)) //添加表达式
            .returns(returnType);//设置返回值

        return builder.build();
    }
}

File: rxhttp-compiler/src/main/java/com/rxhttp/compiler/ParserAnnotatedClass.kt
package com.rxhttp.compiler

import com.squareup.javapoet.*
import rxhttp.wrapper.annotation.Parser
import java.io.IOException
import java.lang.Deprecated
import java.util.*
import javax.annotation.processing.Filer
import javax.lang.model.element.ElementKind
import javax.lang.model.element.ExecutableElement
import javax.lang.model.element.Modifier
import javax.lang.model.element.TypeElement
import javax.lang.model.type.MirroredTypesException
import javax.lang.model.type.TypeMirror
import kotlin.String
import kotlin.collections.ArrayList
import kotlin.require

class ParserAnnotatedClass {

    private val mElementMap = LinkedHashMap<String, TypeElement>()
    private val mTypeMap = LinkedHashMap<String, List<TypeMirror>>()

    fun add(typeElement: TypeElement) {
        val annotation = typeElement.getAnnotation(Parser::class.java)
        val name: String = annotation.name
        require(name.isNotEmpty()) {
            String.format("methodName() in @%s for class %s is null or empty! that's not allowed",
                Parser::class.java.simpleName, typeElement.qualifiedName.toString())
        }
        try {
            annotation.wrappers
        } catch (e: MirroredTypesException) {
            val typeMirrors = e.typeMirrors
            mTypeMap[name] = typeMirrors
        }
        mElementMap[name] = typeElement
    }

    fun getMethodList(filer: Filer): List<MethodSpec> {
        val t = TypeVariableName.get("T")
        val className = ClassName.get(Class::class.java)
        val classTName = ParameterizedTypeName.get(className, t)

        val listTName = ParameterizedTypeName.get(ClassName.get(List::class.java), t)
        val callName = ClassName.get("okhttp3", "Call")
        val responseName = ClassName.get("okhttp3", "Response")
        val requestName = ClassName.get("okhttp3", "Request")
        val parserName = ClassName.get("rxhttp.wrapper.parse", "Parser")
        val progressName = ClassName.get("rxhttp.wrapper.entity", "Progress")
        val logUtilName = ClassName.get("rxhttp.wrapper.utils", "LogUtil")
        val logTimeName = ClassName.get("rxhttp.wrapper.utils", "LogTime")
        val typeName = TypeName.get(String::class.java)
        val parserTName = ParameterizedTypeName.get(parserName, t)
        val simpleParserName = ClassName.get("rxhttp.wrapper.parse", "SimpleParser")
        val type = ClassName.get("java.lang.reflect", "Type")
        val parameterizedType = ClassName.get("rxhttp.wrapper.entity", "ParameterizedTypeImpl")
        val methodList = ArrayList<MethodSpec>()

        methodList.add(
            MethodSpec.methodBuilder("execute")
                .addModifiers(Modifier.PUBLIC)
                .addException(IOException::class.java)
                .addStatement("return newCall().execute()")
                .returns(responseName)
                .build())

        methodList.add(
            MethodSpec.methodBuilder("execute")
                .addModifiers(Modifier.PUBLIC)
                .addTypeVariable(t)
                .addException(IOException::class.java)
                .addParameter(parserTName, "parser")
                .addStatement("return parser.onParse(execute())")
                .returns(t)
                .build())

        methodList.add(
            MethodSpec.methodBuilder("executeString")
                .addModifiers(Modifier.PUBLIC)
                .addException(IOException::class.java)
                .addStatement("return executeClass(String.class)")
                .returns(typeName)
                .build())

        methodList.add(
            MethodSpec.methodBuilder("executeList")
                .addModifiers(Modifier.PUBLIC)
                .addTypeVariable(t)
                .addException(IOException::class.java)
                .addParameter(classTName, "type")
                .addStatement("\$T tTypeList = \$T.get(List.class, type)", type, parameterizedType)
                .addStatement("return execute(new \$T<List<T>>(tTypeList))", simpleParserName)
                .returns(listTName)
                .build())

        methodList.add(
            MethodSpec.methodBuilder("executeClass")
                .addModifiers(Modifier.PUBLIC)
                .addTypeVariable(t)
                .addException(IOException::class.java)
                .addParameter(classTName, "type")
                .addStatement("return execute(new \$T<T>(type))", simpleParserName)
                .returns(t)
                .build())

        methodList.add(
            MethodSpec.methodBuilder("newCall")
                .addAnnotation(Override::class.java)
                .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
                .addCode("""
                    Request request = buildRequest();
                    OkHttpClient okClient = getOkHttpClient();
                    return okClient.newCall(request);
                """.trimIndent())
                .returns(callName)
                .build())

        methodList.add(
            MethodSpec.methodBuilder("buildRequest")
                .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
                .addCode(
                    """
                    boolean debug = ${"$"}T.isDebug();    
                    if (request == null) {
                        doOnStart();
                        request = param.buildRequest();
                        if (debug) 
                            LogUtil.log(request, getOkHttpClient().cookieJar());
                    }
                    if (debug) {
                        request = request.newBuilder()
                            .tag(LogTime.class, new ${"$"}T())
                            .build();
                    }
                    return request;
                """.trimIndent(), logUtilName, logTimeName)
                .returns(requestName)
                .build())

        methodList.add(
            MethodSpec.methodBuilder("doOnStart")
                .addModifiers(Modifier.PRIVATE, Modifier.FINAL)
                .addJavadoc("请求开始前内部调用，用于添加默认域名等操作\n")
                .addStatement("setConverter(param)")
                .addStatement("addDefaultDomainIfAbsent(param)")
                .build())

        if (isDependenceRxJava()) {
            val schedulerName = getClassName("Scheduler")
            val observableName = getClassName("Observable")
            val consumerName = getClassName("Consumer")

            methodList.add(
                MethodSpec.methodBuilder("subscribeOnCurrent")
                    .addAnnotation(Deprecated::class.java)
                    .addJavadoc("@deprecated please user {@link #setSync()} instead\n")
                    .addModifiers(Modifier.PUBLIC)
                    .addStatement("return setSync()")
                    .returns(r)
                    .build())

            methodList.add(
                MethodSpec.methodBuilder("setSync")
                    .addJavadoc("sync request \n")
                    .addModifiers(Modifier.PUBLIC)
                    .addStatement("isAsync = false")
                    .addStatement("return (R)this")
                    .returns(r)
                    .build())

            val observableTName = ParameterizedTypeName.get(observableName, t)
            val consumerProgressName = ParameterizedTypeName.get(consumerName, progressName)

            methodList.add(
                MethodSpec.methodBuilder("asParser")
                    .addModifiers(Modifier.PUBLIC)
                    .addTypeVariable(t)
                    .addParameter(parserTName, "parser")
                    .addParameter(schedulerName, "scheduler")
                    .addParameter(consumerProgressName, "progressConsumer")
                    .addCode("""
                        ObservableCall observableCall;                                      
                        if (isAsync) {                                                      
                          observableCall = new ObservableCallEnqueue(this);                 
                        } else {                                                            
                          observableCall = new ObservableCallExecute(this);                 
                        }                                                                   
                        return observableCall.asParser(parser, scheduler, progressConsumer);
                    """.trimIndent())
                    .returns(observableTName)
                    .build())
        }


        val rxHttpExtensions = RxHttpExtensions()

        //获取自定义的解析器
        for ((parserAlias, typeElement) in mElementMap) {

            //生成RxHttp扩展方法(kotlin编写的方法)
            rxHttpExtensions.generateRxHttpExtendFun(typeElement, parserAlias)

            if (isDependenceRxJava()) { //依赖了RxJava，则生成自定义的asXxx方法
                //onParser方法返回类型
                val returnTypeMirror = getOnParserFunReturnType(typeElement) ?: continue
                val onParserFunReturnType = TypeName.get(returnTypeMirror)

                val typeVariableNames = ArrayList<TypeVariableName>()
                typeElement.typeParameters.forEach {
                    typeVariableNames.add(TypeVariableName.get(it))
                }

                //遍历public构造方法
                getConstructorFun(typeElement).forEach {
                    //根据构造方法参数，获取asXxx方法需要的参数
                    val parameterList = ArrayList<ParameterSpec>()
                    var typeIndex = 0
                    it.parameters.forEach { variableNames ->
                        if (variableNames.asType().toString() == "java.lang.reflect.Type"
                            && typeIndex < typeVariableNames.size
                        ) {
                            //Type类型参数转Class<T>类型
                            val classTypeName = ParameterizedTypeName.get(
                                className, typeVariableNames[typeIndex++])
                            val parameterSpec = ParameterSpec
                                .builder(classTypeName, variableNames.simpleName.toString())
                                .build()
                            parameterList.add(parameterSpec)
                        } else {
                            parameterList.add(ParameterSpec.get(variableNames))
                        }
                    }

                    //方法名
                    var methodName = "as$parserAlias"
                    //方法体
                    val methodBody =
                        "return asParser(new \$T${getTypeVariableString(typeVariableNames)}(${getParamsName(parameterList)}))"

                    //生成的as方法返回类型(Observable<T>类型)
                    var asFunReturnType = ParameterizedTypeName.get(
                        getClassName("Observable"), onParserFunReturnType)
                    methodList.add(
                        MethodSpec.methodBuilder(methodName)
                            .addModifiers(Modifier.PUBLIC)
                            .addTypeVariables(typeVariableNames)
                            .addParameters(parameterList)
                            .addStatement(methodBody, ClassName.get(typeElement))  //方法里面的表达式
                            .returns(asFunReturnType)
                            .build())

                    var haveClassTypeParam = false
                    parameterList.forEach { p ->
                        if (p.type.toString().startsWith("java.lang.Class")) {
                            haveClassTypeParam = true
                        }
                    }
                    //有泛型且有Class类型参数
                    if (typeVariableNames.isNotEmpty() && haveClassTypeParam) {

                        val wrapperListClass: MutableList<ClassName> = mutableListOf()
                        wrapperListClass.add(ClassName.get("java.util", "List"))

                        //泛型的包裹类型，取自Parser注解的wrappers字段
                        mTypeMap[parserAlias]?.forEach { mirror ->
                            val tempClassName = ClassName.bestGuess(mirror.toString())
                            if (!wrapperListClass.contains(tempClassName)) {
                                wrapperListClass.add(tempClassName)
                            }
                        }

                        wrapperListClass.forEach { wrapperClass ->

                            //1、asXxx方法返回值
                            val onParserFunReturnWrapperType = if (onParserFunReturnType is ParameterizedTypeName) {
                                //返回类型有n个泛型，需要对每个泛型再次包装
                                val typeNames = ArrayList<TypeName>()
                                for (typeArg in onParserFunReturnType.typeArguments) {
                                    typeNames.add(ParameterizedTypeName.get(wrapperClass, typeArg))
                                }
                                ParameterizedTypeName.get(onParserFunReturnType.rawType, *typeNames.toTypedArray())
                            } else {
                                ParameterizedTypeName.get(wrapperClass, onParserFunReturnType)
                            }
                            asFunReturnType = ParameterizedTypeName.get(getClassName("Observable"), onParserFunReturnWrapperType)

                            //2、asXxx方法名
                            val name = wrapperClass.toString()
                            val simpleName = name.substring(name.lastIndexOf(".") + 1)
                            methodName = "as$parserAlias${simpleName}"

                            //3、asXxx方法体
                            val funBody = CodeBlock.builder()
                            val paramsName = StringBuilder()
                            //遍历参数，取出参数名
                            parameterList.forEach { param ->
                                if (param.type.toString().startsWith("java.lang.Class")) {
                                    /*
                                     * Class类型参数，需要进行再次包装，最后再取参数名
                                     * 格式：Type tTypeList = ParameterizedTypeImpl.get(List.class, tType);
                                     */
                                    val expression = "\$T ${param.name}$simpleName = \$T.get($simpleName.class, ${param.name})"
                                    funBody.addStatement(expression, type, parameterizedType)
                                    paramsName.append("${param.name}${simpleName}")
                                } else {
                                    paramsName.append(param.name)
                                }
                                paramsName.append(", ")
                            }
                            paramsName.delete(paramsName.length - 2, paramsName.length)
                            val returnStatement = "return asParser(new \$T${getTypeVariableString(typeVariableNames, wrapperClass)}($paramsName))"
                            funBody.addStatement(returnStatement, ClassName.get(typeElement))

                            //4、生成asXxx方法
                            methodList.add(
                                MethodSpec.methodBuilder(methodName)
                                    .addModifiers(Modifier.PUBLIC)
                                    .addTypeVariables(typeVariableNames)
                                    .addParameters(parameterList)
                                    .addCode(funBody.build())  //方法里面的表达式
                                    .returns(asFunReturnType)
                                    .build())
                        }
                    }
                }
            }
        }
        rxHttpExtensions.generateClassFile(filer)
        return methodList
    }


    //获取构造方法
    private fun getConstructorFun(typeElement: TypeElement): MutableList<ExecutableElement> {
        val funList = ArrayList<ExecutableElement>()
        typeElement.enclosedElements.forEach {
            if (it is ExecutableElement
                && it.kind == ElementKind.CONSTRUCTOR
                && it.getModifiers().contains(Modifier.PUBLIC)
            ) {
                funList.add(it)
            }
        }
        return funList
    }

    private fun getParamsName(variableElements: MutableList<ParameterSpec>): String {
        val paramsName = StringBuilder()
        for ((index, element) in variableElements.withIndex()) {
            if (index > 0) paramsName.append(", ")
            paramsName.append(element.name)
        }
        return paramsName.toString()
    }

    //获取泛型字符串 比如:<T> 、<K,V>等等
    private fun getTypeVariableString(typeVariableNames: ArrayList<TypeVariableName>, wrapperClass: ClassName): String {
        val name = wrapperClass.toString()
        val simpleName = name.substring(name.lastIndexOf(".") + 1)

        val type = StringBuilder()
        val size = typeVariableNames.size
        for (i in typeVariableNames.indices) {
            if (i == 0) type.append("<")
            type.append("$simpleName<")
            type.append("${typeVariableNames[i].name}>")
            type.append(if (i < size - 1) "," else ">")
        }
        return type.toString()
    }

    //获取泛型字符串 比如:<T> 、<K,V>等等
    private fun getTypeVariableString(typeVariableNames: ArrayList<TypeVariableName>): String {
        val type = StringBuilder()
        val size = typeVariableNames.size
        for (i in typeVariableNames.indices) {
            if (i == 0) type.append("<")
            type.append(typeVariableNames[i].name)
            type.append(if (i < size - 1) "," else ">")
        }
        return type.toString()
    }


    //获取onParser方法返回类型
    private fun getOnParserFunReturnType(typeElement: TypeElement): TypeMirror? {
        typeElement.enclosedElements.forEach {
            if (it is ExecutableElement   //是方法
                && it.getModifiers().contains(Modifier.PUBLIC)  //public修饰
                && !it.getModifiers().contains(Modifier.STATIC) //非静态
                && it.simpleName.toString() == "onParse"  //onParse方法
                && it.parameters.size == 1  //只有一个参数
                && it.parameters[0].asType().toString() == "okhttp3.Response"  //参数是okhttp3.Response类型
            ) {
                return it.returnType;
            }
        }
        return null
    }
}